### PR TITLE
fix casting to StableId bug in identifier encoding

### PR DESCRIPTION
**Pull request by daesunp**
## Description

This PR fixes a bug where we are incorrectly casting a cursor value to `StableId`.

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/chunked-forest/codec/nodeShape.ts [hunk 1]
```diff
@@ -51,14 +51,17 @@ export class NodeShape extends Shape<EncodedChunkShape> implements NodeEncoder {
...
-			const sessionSpaceCompressedId = cache.idCompressor.tryRecompress(
-				cursor.value as StableId,
-			);
-			const opSpaceCompressedId =
-				sessionSpaceCompressedId !== undefined
-					? cache.idCompressor.normalizeToOpSpace(sessionSpaceCompressedId)
-					: cursor.value;
-			encodeValue(opSpaceCompressedId, this.value, outputBuffer);
+			assert(typeof cursor.value === "string", "identifier must be type string");
+			if (!isStableId(cursor.value)) {
```
Comment (noencke) [inline packages/dds/tree/src/feature-libraries/chunked-forest/codec/nodeShape.ts:55]: nit: for cases where there is one "success" path and multiple "fallback" paths, either structure it like this:
```ts
if (conditionA && conditionB) {
	// Success
} else {
	// Failure
}
```
or, in a helper function with returns like this:
```ts
{
	if (conditionA) {
		if (conditionB) {
			return success;
		}
	}
	return failure;
}
```

The current code is structured like:

```ts
if (!conditionA) {
	// Failure
} else {
	if (conditionB) {
		// Success
	} else {
		// Failure
	}
}
```

- note how there are two identical "failure" cases - i.e. the cases are not "factored out" in the algebraic sense. There are times when it's fine to have multiple cases too, like this, which reduces nesting:

```ts
{
	if (!conditionA) {
		return false;
	}

	if (!conditionB) {
		return false;
	}

	return true;
}
```

But all the "failure" cases are at least still together.

File: packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkEncodingEndToEnd.spec.ts [hunk 2]
```diff
@@ -280,5 +280,31 @@ describe("End to end chunked encoding", () => {
 			// Check that the identifierValue is the original uncompressed id.
 			assert.equal(identifierValue, id);
 		});
+
+		it("is the uncompressed value when it is an invalid uuid", () => {
```
Comment (noencke) [inline packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkEncodingEndToEnd.spec.ts:284]:
```suggestion
		it("is the uncompressed value when it is not a UUID", () => {
```

File: packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkEncodingEndToEnd.spec.ts [hunk 3]
```diff
@@ -280,5 +280,31 @@ describe("End to end chunked encoding", () => {
...
+			// generate an id from a different id compressor.
```
Comment (noencke) [inline packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkEncodingEndToEnd.spec.ts:285]: Remove this comment

**Review by taylorsw04 — APPROVED** [review]

File: packages/dds/tree/src/feature-libraries/chunked-forest/codec/nodeShape.ts [hunk 4]
```diff
@@ -39,6 +40,24 @@ export class NodeShape extends Shape<EncodedChunkShape> implements NodeEncoder {
...
 
+	private getEncodeValue(cursor: ITreeCursorSynchronous, cache: EncoderCache): Value {
+		if (this.value === 0) {
+			assert(typeof cursor.value === "string", "identifier must be type string");
+			if (
+				isStableId(cursor.value) &&
+				cache.idCompressor.tryRecompress(cursor.value) !== undefined
+			) {
+				const sessionSpaceCompressedId = cache.idCompressor.tryRecompress(cursor.value);
+				assert(
```
Comment (noencke) [inline packages/dds/tree/src/feature-libraries/chunked-forest/codec/nodeShape.ts:51]: Is this correct? We weren't asserting before. What happens if you pass in a string that was not generated by the compressor, but is a UUID? Like `b3767cd8-3dfa-451b-83d8-36571ad6b72b`? In that case, we want to fallback to the default case (return cursor.value), not fail - right? If that's true, there should be another unit test added for that case as well.

File: packages/dds/tree/src/feature-libraries/chunked-forest/codec/nodeShape.ts [hunk 5]
```diff
@@ -39,6 +40,24 @@ export class NodeShape extends Shape<EncodedChunkShape> implements NodeEncoder {
 		this.explicitKeys = new Set(this.fields.map((f) => f.key));
 	}
 
+	private getEncodeValue(cursor: ITreeCursorSynchronous, cache: EncoderCache): Value {
```
Comment (noencke) [inline packages/dds/tree/src/feature-libraries/chunked-forest/codec/nodeShape.ts:43]:
```suggestion
	private getValueToEncode(cursor: ITreeCursorSynchronous, cache: EncoderCache): Value {
```

File: packages/dds/tree/src/feature-libraries/chunked-forest/codec/nodeShape.ts (hunk 4 above)
Comment (noencke) [reply packages/dds/tree/src/feature-libraries/chunked-forest/codec/nodeShape.ts:51]: Oh - nevermind, I see you are calling `tryRecompress` twice. Instead, just make it another `if` conditional and only do it once.

**Review by noencke — APPROVED** [review]

**Pull request merged**